### PR TITLE
Always explicitly restart, even if the service isn't running, to propagate that action to dependent services.

### LINKED ISSUE #77337: systemd state=restarted only starts - not restarts - a stopped service, affecting dependent services
### Summary

I use the `systemd` module and `state: restarted` to always restart a service and all its dependent services.
This works well when the service in question is running: A restart signal is sent to that service and propagated to all its dependent services.
However, systemd translates a "restart" request for a stopped service to a "start" request:

https://github.com/ansible/ansible/blob/14098d9b045f4048e540560d3bee808cbb87e2fa/lib/ansible/modules/systemd.py#L546-L549

If the service in question is currently stopped, Ansible will only ask it to start. If the dependent services are already running, they won't be affected at all. This is wrong as they should be restarted.

If I ran `systemctl restart MY_SERVICE` instead, my service and all dependent services would have been restarted, no matter if they were previously running or not.

IMO, this check is just wrong and can be simply removed to fix my problem.
This is similar to #30076, but I get why you can't just use "restart" for the `service` module if you want to stay LSB-compatible. This should be no reason for the `systemd` module though. It always supports restarting.

### Issue Type

Bug Report

### Component Name

systemd

### Ansible Version

```console
$ ansible --version
ansible [core 2.12.2]
  config file = /srv/ansible/repo/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.8.10 (default, Nov 26 2021, 20:14:08) [GCC 9.3.0]
  jinja version = 2.10.1
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed
```

### OS / Environment

Ubuntu 20.04 LTS

### Steps to Reproduce

The relevant line in my playbook is:
```yaml (paste below)
- name: "Restart MY_SERVICE"
  service:
    name: "MY_SERVICE"
    state: "restarted"
```

The target system has MY_SERVICE and MY_OTHER_SERVICE, with MY_OTHER_SERVICE being a dependency of MY_SERVICE via `/etc/systemd/system/MY_OTHER_SERVICE.service.d/restart_after_MY_SERVICE.conf`:

```
[Unit]
PartOf=MY_SERVICE.service
After=MY_SERVICE.service
```

### Expected Results

I expected the same result as `systemctl restart MY_SERVICE`, namely that both MY_SERVICE and MY_OTHER_SERVICE are restarted, no matter their previous state.

### Actual Results

```console
If MY_SERVICE is running, things work as expected.
If MY_SERVICE is stopped and MY_OTHER_SERVICE is running, MY_SERVICE is started, but MY_OTHER_SERVICE is not affected at all.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by ColinFinck**
##### SUMMARY
Fixes #77337

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
systemd

### CODE REVIEW

**Comment by ColinFinck** [comment]
Bump

Who can review this?

File: lib/ansible/modules/systemd.py [hunk 1]
```diff
@@ -542,11 +542,16 @@ def main():
...
+                    # Always explicitly restart, even if the service isn't running, to propagate that action to dependent services.
+                    # https://github.com/ansible/ansible/issues/77337
+                    action = 'restart'
+                    result['state'] = 'started'
+                else:  # 'reloaded'
                     if not is_running_service(result['status']):
                         action = 'start'
                     else:
-                        action = module.params['state'][:-2]  # remove 'ed' from restarted/reloaded
+                        action = 'reload'
```
Comment (bcoca) [inline lib/ansible/modules/systemd.py:554]: This PR is missing a changelog and tests, but also I don't think this is the correct approach. This module has the same problem with other commands (dependencies are not accounted for) and requires a more generic solution to handle all cases by always taking dependencies into consideration.

Sadly I have not found a good way to get this information in a usable form from systemd w/o going into dbus and that makes the solution very complicated.

https://docs.ansible.com/ansible/devel/community/development_process.html#micro-development-the-lifecycle-of-a-pr

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/systemd.py (hunk 1 above)
Comment (ColinFinck) [reply lib/ansible/modules/systemd.py:554]: I've added a changelog fragment as requested. I could also add a test if you could point me to an existing systemd test for restarting a service.

While my approach may not solve all problems with dependencies, it solves the specific problem I laid out in #77337.
As you write yourself that a generic solution to solve all systemd dependency problems is likely out of reach for the foreseeable future, I would appreciate if my PR could be merged as a first step.

**Comment by bcoca** [comment]
https://github.com/ansible/ansible/tree/devel/test/integration/targets/systemd

**Comment by bcoca** [comment]
closing due to inactivity, linked comment for new approach on original ticket

**Pull request closed without merging**